### PR TITLE
Remove SourceFile::SynthesizedDecls

### PR DESCRIPTION
**Pull request by CodaFi**
This list used to matter a lot more when `typeCheckFunctionsAndExternalDecls` actually did what it said and dealt with external functions.  Nowadays it's just dead weight.  Drop it and clean up after it.  We now eagerly typecheck derived declarations as soon as they're inserted into their parent decl context.  The final commit also works around Hashable not quite having the smarts to pre-reject invalid derivation contexts so we don't emit a "Hashable is broken" message.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
I doubt this will have compatibility impacts, but it changes validation order so.

@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
⛵️

File: lib/Sema/DerivedConformances.cpp [hunk 1]
```diff
@@ -35,12 +35,16 @@ DeclContext *DerivedConformance::getConformanceContext() const {
...
 }
 
-void DerivedConformance::addMembersToConformanceContext(
+bool DerivedConformance::addMembersToConformanceContext(
     ArrayRef<Decl *> children) {
   auto IDC = cast<IterableDeclContext>(ConformanceDecl);
+  bool anyInvalid = false;
   for (auto child : children) {
     IDC->addMember(child);
+    TypeChecker::typeCheckDecl(child);
```
Comment (slavapestov) [inline lib/Sema/DerivedConformances.cpp]: This is going to do extra work in secondary files and slow down compilation. Instead, you should check conformances before visiting members in the DeclChecker, then the ordinary member walk in `visit(Class|Struct|Enum)Decl()` will handle them.

File: lib/Sema/TypeCheckDecl.cpp [hunk 2]
```diff
@@ -3675,6 +3675,7 @@ void TypeChecker::typeCheckDecl(Decl *D) {
 // Returns 'nullptr' if this is the setter's 'newValue' parameter;
 // otherwise, returns the corresponding parameter of the subscript
 // declaration.
+
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: Is this an accidental change?

File: lib/Sema/TypeChecker.cpp [hunk 3]
```diff
@@ -294,7 +294,6 @@ static void bindExtensions(SourceFile &SF) {
 
 static void typeCheckFunctionsAndExternalDecls(SourceFile &SF, TypeChecker &TC) {
```
Comment (slavapestov) [inline lib/Sema/TypeChecker.cpp]: Looks like 'SF' isn't used anywhere in the body, so we can remove it from the parameter list

Comment (CodaFi) [reply lib/Sema/TypeChecker.cpp]: I was hoping to remove this function altogether in #28138

Comment (CodaFi) [reply lib/Sema/TypeChecker.cpp]: I was hoping to get rid of this function altogether in #28138.

**Pull request merged**